pycbc/types/timeseries.py

```diff
@@ -595,15 +595,34 @@ def gate(self, time, window=0.25, method='taper', copy=True,
             # Uses the hole-filling method of
             # https://arxiv.org/pdf/1908.05644.pdf
             from pycbc.strain.gate import gate_and_paint
+            from pycbc.waveform.utils import apply_fd_time_shift
             if invpsd is None:
                 # These are some bare minimum settings, normally you
                 # should probably provide a psd
                 invpsd = 1. / self.filter_psd(self.duration/32, self.delta_f, 0)
             lindex = int((time - window - self.start_time) / self.delta_t)
-            rindex = lindex + int(2 * window / self.delta_t)
+            rindex = int((time + window - self.start_time) / self.delta_t)
             lindex = lindex if lindex >= 0 else 0
             rindex = rindex if rindex <= len(self) else len(self)
-            return gate_and_paint(data, lindex, rindex, invpsd, copy=False)
+            rindex_time = data.sample_times[rindex]
+            offset = rindex_time - (time + window)
+            if offset == 0:
+                return gate_and_paint(data, lindex, rindex, invpsd, copy=False)
+            else:
+                # time shift such that gate end time lands on a specific data sample
+                fdata = data.to_frequencyseries()
+                fdata = apply_fd_time_shift(fdata, offset + fdata.epoch, copy=False)
```

Comment:
Have you double checked the sign on these things? I think the logic makes sense, but I always screw up signs.

Reply:
I'm careful to take `offset` less than 0 and equal to the difference between the genuine gate_end_time and the nearest (floor) discrete sample's time. The reasoning can be explained as that the previous codes always inpaint data until earlier than we want. Putting another way, although we want to inpaint until gate_end_time, it only stops until a floor time. Therefore we shift the data back to earlier, such that more later data can be inpainted. I hope this understanding makes sense.